### PR TITLE
Refactor channel call management

### PR DESCRIPTION
**Pull request by nathanielmanistaatgoogle**
This is prerequisite to @ncteisen's [invalid metadata handling](https://github.com/grpc/grpc/issues/8814).

Including @ctiller for the `grpc.h` documentation change.

### CODE REVIEW

File: src/python/grpcio/grpc/_channel.py [hunk 1]
```diff
@@ -358,7 +358,7 @@ def add_callback(self, callback):
       if self._state.callbacks is None:
         return False
       else:
-        self._state.callbacks.append(lambda: callback())
+        self._state.callbacks.append(callback)
```
Comment (kpayson64) [inline src/python/grpcio/grpc/_channel.py]: +1

**Comment by ncteisen** [comment]
Looks good, thanks for the refactor!

File: src/python/grpcio/grpc/_channel.py (hunk 1 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_channel.py]: [Ah, memories...](https://github.com/PyCQA/pylint/commit/23fd11df220ae230ba20362b8fd94d3b03b695c0)

**Comment by nathanielmanistaatgoogle** [comment]
I have secured @ctiller's approval for this offline; this will be merged after the test fixit is declared over.

**Pull request merged**